gears/data_utils.py

```diff
@@ -3,12 +3,83 @@
 import numpy as np
 import pandas as pd
 import scanpy as sc
+from scipy.sparse import csr_matrix
 
 from .utils import parse_any_pert
 
 sc.settings.verbosity = 0
 warnings.filterwarnings("ignore")
 
+def update_adata_for_gears(adata, control_name, cell_type):
+    """
+    Updates adata to required formatting for
+    GEARS PertData.new_data_process() method
+
+    Parameters
+    ----------
+    adata : AnnData
+        AnnData object containing raw data
+    control_name : str
+        Name of control perturbation in adata.obs.condition
+    cell_type : str
+        Name of cell type used in experiment
+    """
+
+    # Specify gene_name in var and cell_type in obs
+    adata.var['gene_name'] = adata.var.index.values
+    adata.obs['cell_type'] = cell_type
+
+    # Change name of control perturbation to 'ctrl'
+    adata.obs.condition = adata.obs.condition.astype(str)
+    for i in range(adata.n_obs):    
+        if adata.obs.condition.iloc[i] == control_name:
+            adata.obs.condition.iloc[i] = 'ctrl'
+    adata.obs.condition = adata.obs.condition.astype('category')
+
+    # Remove unnecessary columns from obs and var to
+    # solve weird & opaque errors
+    adata.obs = adata.obs[['condition','cell_type']]
+    adata.var = adata.var[['gene_name']]
+
+    # If adata.X is a numpy array, convert to csr_matrix
+    if isinstance(adata.X, np.ndarray):
+        adata.X = csr_matrix(adata.X)
+
+    return adata
+
+
+def add_in_missing_genes_for_perturbations(adata):
+    """ GEARs requires that all perturbations have a corresponding gene
+    in adata.var. For perturbations with no gene, this function adds a
+    vector of zeros to adata.X for that gene and adds the gene to adata.var.
+    """
+
+    adata.var = adata.var.drop(columns=adata.var.columns.values)
+    perturbations = []
+    conditions = adata.obs.condition.unique()
+    for cond in conditions:
+        for p in cond.split("+"):
```

Comment:
Maybe make the delimiter an argument, since we'll likely see datasets where the delimiter is not '+'

Reply:
yup yup